app/src/main/java/org/mozilla/tv/firefox/tabs/TabsAdapter.kt

```diff
@@ -0,0 +1,106 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.tabs
+
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import mozilla.components.browser.session.Session
+import mozilla.components.concept.tabstray.TabsTray
+import mozilla.components.support.base.observer.Observable
+import mozilla.components.support.base.observer.ObserverRegistry
+import org.mozilla.tv.firefox.R
+
+const val ADD_VIEW_TYPE = 300
+const val SESSION_VIEW_TYPE = 200
+
+/**
+ * RecyclerView adapter implementation to display a list/grid of tabs.
+ */
+@Suppress("TooManyFunctions")
+class TabsAdapter(
+    delegate: Observable<TabsTray.Observer> = ObserverRegistry()
+) : RecyclerView.Adapter<RecyclerView.ViewHolder>(),
+    TabsTray,
+    Observable<TabsTray.Observer> by delegate {
+
+    internal lateinit var tabsTray: WebRenderTabsTray
+
+    private val holders = mutableListOf<TabViewHolder>()
+    private var sessions: List<Session> = listOf()
+    private var selectedIndex: Int = -1
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+        if (viewType == ADD_VIEW_TYPE) {
+            return TabPlusHolder(
+                    LayoutInflater.from(parent.context).inflate(
+                            R.layout.tabstray_add_item,
+                            parent,
+                            false),
+                    tabsTray
+            )
+        } else {
```

Comment:
I would prefer if we checked both conditions, then threw an exception in the else branch.  If we ever add a new type to this RV, it would be easy to miss one of these sections and cause strange bugs.